src/components/modal-container.js

```diff
@@ -134,101 +145,54 @@ export default function ModalContainerFactory(
-        downloadFile(file, DEFAULT_EXPORT_IMAGE_NAME);
+      if (!this.props.uiState.exportImage.exporting) {
+        exportImage(this.props, this.props.uiState.exportImage);
+        this.props.uiStateActions.cleanupExportImage();
+        this._closeModal();
       }
-      this.props.uiStateActions.cleanupExportImage();
-      this._closeModal();
     };
 
-    _downloadFile(data, type, filename) {
-      const fileBlob = new Blob([data], {type});
-      downloadFile(fileBlob, filename);
-    }
-
     _onExportData = () => {
-      const {visState, uiState} = this.props;
-      const {datasets} = visState;
-      const {selectedDataset, dataType, filtered} = uiState.exportData;
-      // get the selected data
-      const filename = 'kepler-gl';
-      const selectedDatasets = datasets[selectedDataset] ?
-        [datasets[selectedDataset]] : Object.values(datasets);
-
-      if (!selectedDatasets.length) {
-        // error: selected dataset not found.
-        this._closeModal();
-      }
-
-      selectedDatasets.forEach(selectedData => {
-        const {allData, fields, label, filteredIdxCPU = []} = selectedData;
-        const exportData = filtered ? filteredIdxCPU.map(i => allData[i]) : allData;
-        // start to export data according to selected data type
-        switch (dataType) {
-          case EXPORT_DATA_TYPE.CSV: {
-            const type = 'text/csv';
-            const csv = formatCsv(exportData, fields);
-            this._downloadFile(csv, type, `${filename}_${label}.csv`);
-            break;
-          }
-          // TODO: support more file types.
-          default:
-            break;
-        }
-      });
-
+      exportData(this.props, this.props.uiState.exportData);
       this._closeModal();
     };
 
-    _onExportJSONMap = () => {
+    _onExportMap = () => {
       const {uiState} = this.props;
-      const {hasData} = uiState.exportMap[EXPORT_MAP_FORMATS.JSON];
-
-      // we pass all props because we avoid to create new variables
-      const data = hasData ? KeplerGlSchema.save(this.props)
-        : KeplerGlSchema.getConfigToSave(this.props);
-
-      this._downloadFile(
-        JSON.stringify(data, null, 2),
-        'application/json',
-        'keplergl.json'
+      const {format} = uiState.exportMap;
+      (format === EXPORT_MAP_FORMATS.HTML ? exportHtml : exportJson)(
+        this.props,
+        this.props.uiState.exportMap[format] || {}
       );
-
       this._closeModal();
     };
 
-    _onExportHTMLMap = () => {
-      const {uiState} = this.props;
-      const {userMapboxToken, exportMapboxAccessToken, mode} = uiState.exportMap[EXPORT_MAP_FORMATS.HTML];
+    _exportFileToCloud = ({provider, isPublic, closeModal}) => {
+      const toSave = exportMap(this.props);
+      this.props.providerActions.exportFileToCloud({
+        mapData: toSave,
+        provider,
+        isPublic,
+        closeModal,
+        onSuccess: this.props.onExportToCloudSuccess,
```

Comment:
we are using these two callbacks (onSuccess, onError) from within redux using trunk.

Is this a better pattern than using redux to set attributes in the store and upon detecting new values in the view we call onSuccess or onError?

Reply:
`onSuccess` and `onError` are action handlers. kepler.gl core uses tasks, instead of thunk. it creates a task that dispatches the custom actions if user-defined

Reply:
got it.